src/box/txn_limbo.c

```diff
@@ -671,41 +763,30 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
 		return;
 	assert(limbo->owner_id != REPLICA_ID_NIL);
 	int64_t prev_lsn = vclock_get(&limbo->vclock, replica_id);
+	assert(lsn >= prev_lsn);
 	/*
 	 * One of the reasons why can happen - the remote instance is not
-	 * read-only and wrote something under its own insance_id. For qsync
+	 * read-only and wrote something under its own instance_id. For qsync
 	 * that most likely means that the remote instance decided to take over
 	 * the limbo ownership, and the current node is going to become a
 	 * replica very soon.
 	 */
 	if (lsn == prev_lsn)
 		return;
 	vclock_follow(&limbo->vclock, replica_id, lsn);
-	struct txn_limbo_entry *e;
-	int64_t confirm_lsn = -1;
-	rlist_foreach_entry(e, &limbo->queue, in_queue) {
-		assert(e->ack_count <= VCLOCK_MAX);
-		if (e->lsn > lsn)
-			break;
-		/*
-		 * Sync transactions need to collect acks. Async
-		 * transactions are automatically committed right
-		 * after all the previous sync transactions are.
-		 */
-		if (!txn_has_flag(e->txn, TXN_WAIT_ACK)) {
-			continue;
-		} else if (e->lsn <= prev_lsn) {
-			continue;
-		} else if (++e->ack_count < replication_synchro_quorum) {
-			continue;
-		} else {
-			confirm_lsn = e->lsn;
-		}
+
+	if (limbo->confirmed_entry_is_valid) {
+		struct txn_limbo_entry *e = rlist_next_entry(
+			limbo->confirmed_entry, in_queue);
+		/* Make sure confirmed_entry is valid. It must be true. */
+		assert(!rlist_entry_is_head(e, limbo->queue, in_queue));
+		assert(txn_has_flag(e->txn, TXN_WAIT_ACK));
+		assert(e->lsn != -1);
+		if (e->lsn <= prev_lsn || lsn < e->lsn)
+			return;
+		++limbo->ack_count;
 	}
-	if (confirm_lsn == -1 || confirm_lsn <= limbo->confirmed_lsn)
-		return;
-	txn_limbo_write_confirm(limbo, confirm_lsn);
-	txn_limbo_read_confirm(limbo, confirm_lsn);
+	txn_limbo_confirm(limbo);
```

Comment:
Why do you call confirm even when `limbo->ack_count` didn't change?

Reply:
This does not mean that `ack_count` has not changed. If `confirmed_entry_is_valid == false`, this means that inside the `txn_limbo_confirm` call, we will try to make `confirmed_entry` valid again and calculate the correct `ack_count` value using `vclock_count_ge`. This ACK will be taken into account since before calling `txn_limbo_confirm` we executed `vclock_follow(&limbo->vclock, replica_id, lsn);` (https://github.com/Astronomax/tarantool/blob/6eee079a537a40c9721f6ca8a9fe616748f3b5c6/src/box/txn_limbo.c#L753C2-L753C49).

Reply:
Yeah, missed that, thanks.